Add CommentPruner to Prow to save API tokens when deleting stale comments.

cjwagner opened the pull request:
This is a prow parallel to mungegithub's comment-deleter munger.

/cc @rmmh

kargakis commented on prow/commentpruner/commentpruner.go:
```diff
@@ -0,0 +1,98 @@
+	comments []github.IssueComment
+
+	once sync.Once  // Used to fetch the IssueComments once.
+	lock sync.Mutex // Protects `comments`.
```
As a convention locks are added on top of the fields they protect.

cjwagner replied:
That makes sense. Fixed.

spxtr commented:
/lint

cjwagner commented:
/retest

spxtr commented:
What plugins need this?

cjwagner commented:
> What plugins need this?

This is for some of the plugins that are being migrated from mungegithub. It will also be used by the release-note plugin.

spxtr commented:
Can you go into more detail about why plugins need this? I'm worried that it will actually cause *more* token usage, because instead of plugins smartly deciding when they need to delete comments, they will instead call `PruneComments` on every webhook or something.

That said, if it makes migrating mungegithub -> prow easier then I'm all for it. It's a well-designed package :)

cjwagner commented:
Plugins should still smartly decide to use `PruneComments` and should not use it on every webhook event. That being said, once all the mungers are migrated to Prow, some webhook events that hook receives will be likely to have multiple handlers request comment pruning. Without this client, the handlers would each have to fetch the comments which would use more API tokens.
Even an HTTP cache shared by the handlers wouldn't completely minimize token usage like this client does because each deletion causes the list of issues on github to change, invalidating the cache entry.

kargakis commented on prow/commentpruner/commentpruner.go:
```diff
@@ -0,0 +1,98 @@
+	}
+}
+
+// PruneComments fetchest issue comments if they have not yet been fetched for this webhook event
```
typo

cjwagner replied:
Fixed.

rmmh commented on prow/commentpruner/commentpruner.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+// bufSize is the size of the sliding window and subsequently the cache size.
+const bufSize = 100
```
unused

cjwagner replied:
Fixed.

rmmh commented on prow/commentpruner/commentpruner.go:
```diff
@@ -0,0 +1,98 @@
+func (c *EventClient) PruneComments(shouldPrune func(github.IssueComment) bool) {
+	c.once.Do(func() {
+		c.lock.Lock()
+		defer c.lock.Unlock()
```
the locking here is unnecessary-- once provides its own lock

kargakis replied:
The lock is there likely because `once` does not protect a second thread from accessing the comment slice.

cjwagner replied:
I think @rmmh  is right. All threads block on `c.once.Do` until the function has returned. See [here](https://golang.org/pkg/sync/#Once.Do).

rmmh commented on prow/hook/events.go:
```diff
@@ -39,6 +40,13 @@ func (s *Server) handleReviewEvent(l *logrus.Entry, re github.ReviewEvent) {
 			pc.Logger = l.WithField("plugin", p)
 			pc.Config = s.ConfigAgent.Config()
 			pc.PluginConfig = s.Plugins.Config()
+			pc.CommentPruner = commentpruner.NewEventClient(
```
this is kind of repetitive but I guess the whole file is

cjwagner commented:
Comments addressed. PTAL

cjwagner commented:
Addressed comments and updated to only consider bot comments for pruning so that handlers don't have to include that check themselves (plugins should never need to delete human made comments).

rmmh commented:
/lgtm
